Release command 🚀

jordisala1991 opened the pull request:
Command to make releases easy.

## ToDo

- [x] Compile changelog
- [x] List PR
- [x] get labels of all PR
- [x] Determine what is the next tag
- [x] Check if the base branch is okay to do a release: build is okay
- [x] Print labels with colors, mark the ones that are not supposed to be there
- [x] Check if changelog is present on all PR that need to have it
- [x] Cleanup code

## Nice to have
- [ ] Move some responsabilities out of the command
- [ ] Add test

## Subject

There are a lot of manual tasks that can be done automatically. Maybe we don't start with a full automatic release process, but this will be the first step.

- [x] Check if we can create a commit with the changelog modifications and create a tag and a release on the repository

Yes, we can:

https://developer.github.com/v3/repos/contents/#get-contents
https://developer.github.com/v3/repos/contents/#update-a-file
https://developer.github.com/v3/git/tags/#create-a-tag-object
https://developer.github.com/v3/repos/releases/#create-a-release

jordisala1991 commented:
I know the code is totally crap right now.

This is some example of output when executing this command against `doctrine-orm-admin-bundle`:

```

sonata-project/doctrine-orm-admin-bundle
========================================

Pull requests
-------------

added captions to TOC tree in docs (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/669
update one to many documentation (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/667
Fixed wrong translation in one_to_many template (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/666
Small clanup (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/662
Allow simplethings/entity-audit-bundle ^1.0 (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/661
Fix check access to object in view (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/660
Fixed duplicate translation of "Delete" in edit tab view (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/658
Translation is broken in oneToMany edit form (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/657
Fixed wrong entity link in list view (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/654
Use the right array conversion for filter value (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/653
Remove deprecated translation logic (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/651
TYPE_IS_NOT_EQUAL doesn't consider NULL values (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/650
TYPE_NOT_CONTAINS doesn't consider NULL values (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/649
Fixed deprecated BaseBlockService usage (#641) (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/648
Apply fixes from StyleCI (patch)
    https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/647


Changelog
---------

### Changed
- translation in twig templates uses the twig translation filter (#651)
### Fixed
- Fixed wrong translation in delete checkbox in `edit_orm_one_to_many_inline_table.html.twig` (#666)
- Fixed duplicate translation of "Delete" in edit tab view (#658)
- Fixed broken translation in oneToMany table view (#657)
- Fixed wrong link generation when calling `DatagridMapper::addIdentifier` on mapped field (#654)
- Use the right array conversion for filter value (#653)
- Consider NULL values when using 'is not equal' advanced model filter (#650)
- Consider NULL values when using 'does not contain' advanced string filter (#649)
- deprecated usage of Sonata\BlockBundle\Block\BaseBlockService (#648)
### Security
- Fixed view - check specific item collection, not to the whole collection. (#660)
```

greg0ire commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,174 @@
+            $this->io->title($package->getName());
+            $this->prepareRelease($package, $branch, $output);
+        } catch (ExceptionInterface $e) {
+            $this->io->error('Failed with message: '.$e->getMessage());
```
What's the point? Also, maybe not return 0?

greg0ire replied:
Let's remove the try catch, it just makes things like bad credentials harder to troubleshoot IMO

jordisala1991 commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,174 @@
+            $this->io->title($package->getName());
+            $this->prepareRelease($package, $branch, $output);
+        } catch (ExceptionInterface $e) {
+            $this->io->error('Failed with message: '.$e->getMessage());
```
Just copied that part from another command of dev-kit. Not sure if it is needed :D

jordisala1991 commented:
Disabled travis for now on this PR, I don't want to trigger a lot of builds when pushing commits

sstok commented:
https://github.com/park-manager/hubkit Feel free to reuse the Version class which automatically determines the next possible version, validates gaps, etc 😃

jordisala1991 commented:
Hmm, that would be nice, but I know what the next version should be by looking at the labels of all PRs. Is there a way to do that with your class @sstok ?

jordisala1991 commented:
I have a question about github.

When you tag a new version, how do you mark it as the latest release?

The thing is, when executing this command, I check for the latest release, but looking at `doctrine-orm-admin-bundle` the latest release is not the latest :trollface: .

https://github.com/sonata-project/SonataDoctrineORMAdminBundle/releases

jordisala1991 commented:
We can use this: https://developer.github.com/v3/repos/contents/#get-contents

To get the content of the changelog, then prepend the new changelog and use that other method:

https://developer.github.com/v3/repos/contents/#update-a-file

To update it and create a commit for the next release, then use this:

https://developer.github.com/v3/git/tags/#create-a-tag-object

To create the tag pointing to that commit sha, and finally use this:

https://developer.github.com/v3/repos/releases/#create-a-release

To create the release pointing to that tag. 🎉 

WDYT @sonata-project/contributors, sounds nice, but also risky if something goes wrong

sstok commented:
> I mean, given a current release and a minor, patch string or integer give me the next release

Yes, you can take a version say v1.2.4 and ask for an increase of any part, including alpha, beta, rc, etc. v1.2.4 + minor = v1.3.0 

https://github.com/park-manager/hubkit/blob/master/src/Helper/Version.php#L177

But there is a minor catch if you want to jump from v1.2.4 to 2.0.0-ALPHA1, you need to create this version explicit as you can also have v1.3.0-ALPHA1.

And all versions must follow the semver scheme (`[v]major.minor.patch[-mataver]`), nightly builds are simple ignored.

I will move these classes to there own package today (or tomorrow 😉 ).

jordisala1991 commented:
Thanks @sstok , will try to refactor the `determineNextVersion` method using your package. 👍

greg0ire commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,289 @@
+ */
+final class ReleaseCommand extends AbstractCommand
+{
+    protected static $labels = array(
```
Protected property on a final class?

jordisala1991 replied:
👍 changed to private

OskarStark commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,289 @@
+    private static $stabilities = array(
+        'patch' => 'blue',
+        'minor' => 'green',
+        'stable' => 'yellow',
```
shouldn't this be `major`

greg0ire replied:
I think Doctrine uses "Improvement". It's nice because it does not feel insulting for the contributor.

OskarStark commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,289 @@
+        $this
+            ->setName('release')
+            ->setDescription('Makes a release of a project.')
+            ->addArgument('project', InputArgument::REQUIRED, 'The project to do a release of.')
```
could you please add a help command with examples?

jordisala1991 replied:
Yes, do you mean a comment with some examples or the class command itself has a method to do that?

OskarStark replied:
<img width="1048" alt="screenshot 2017-03-20 12 22 20" src="https://cloud.githubusercontent.com/assets/995707/24097831/e540a158-0d67-11e7-84e6-8612f6c436c6.png">

Something like this

jordisala1991 replied:
Added, my english is pretty bad, but @greg0ire could help improving the text :P .

jordisala1991 commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,289 @@
+    private static $stabilities = array(
+        'patch' => 'blue',
+        'minor' => 'green',
+        'stable' => 'yellow',
```
I use stable for pedantic changes. I didn't add major because major release will require special work, that maybe cannot be automated, or atleast not in the same way

OskarStark replied:
ok

greg0ire replied:
What's wrong with `pedantic`? `stable` is a bit vague IMO, and `patch` / `minor` releases could happen on a stable branch, so it might be confusing.

jordisala1991 replied:
I would like a word for PRs that does not require a release. We could use pedantic, but then docs PR should get this pedantic status.

Summary:

Given a currentRelease 3.3.0 and a Pull request with the following status:

'patch' => will result in 3.3.1
'minor' => will result in 3.4.0
'????' => will result in 3.3.0 (no release needed)
'major' => will result in 4.0.0 (not the scope of this PR)

What word should I use? @greg0ire @OskarStark

greg0ire replied:
Harmless? No_impact? Irrelevant? Idem potent?

jordisala1991 replied:
Like irrelevant 👍

greg0ire replied:
Unnoticeable? Furtive?

jordisala1991 replied:
![](https://media.giphy.com/media/12m6YBQBVVPMK4/giphy.gif)

greg0ire replied:
Lol

OskarStark commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,304 @@
+<info>php dev-kit release doctrine-orm-admin-bundle</info>
+
+The command will show what is the status of the project, then a list of pull requests
+made against selected branch (default: <info>3.x</info>) with the following information: stability, name,
```
couldn't we parse our projects.yml and mark the default branch as default there?

jordisala1991 replied:
Changed this copy, see comment below

OskarStark commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,304 @@
+        $this
+            ->setName('release')
+            ->setDescription('Helps with a project release.')
+            ->addArgument('project', InputArgument::REQUIRED, 'The project to release.')
```
what about making this also interactive and having autocomplete byes on the `projects.yml` ❓

jordisala1991 replied:
Implemented

OskarStark commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,304 @@
+            ->setName('release')
+            ->setDescription('Helps with a project release.')
+            ->addArgument('project', InputArgument::REQUIRED, 'The project to release.')
+            ->addOption('branch', null, InputOption::VALUE_OPTIONAL, 'branch to release', '3.x')
```
same autocompletion here?

jordisala1991 replied:
I Removed this parameter, not the branch detection is done based on projects.yml

jordisala1991 commented:
My proposal is to leave this PR with only a command to show what the current state of the branch and what will be the changelog of the new version, status of merged PRs, etc..

And create another PR after this one is merged to do the automatic Pull request with the new changelog (maybe with automerge) and show to the user what commands needs to execute to create the tag (maybe?)

WDYT @sonata-project/contributors

OskarStark commented:
I executed it for the `translation-bundle` and got this result:
![screenshot 2017-03-22 10 16 57](https://cloud.githubusercontent.com/assets/995707/24190326/b39e31e2-0ee8-11e7-88dd-9abcedbff460.png)

totally fine, except one thing:
![screenshot 2017-03-22 10 16 18](https://cloud.githubusercontent.com/assets/995707/24190345/bd2f3a12-0ee8-11e7-84c8-d6ab5eb1e961.png)

This is **not** `IRRELEVANT`, there is only the `require review` label. We should verify, that at least one of `docs, pedantic, patch, minor` is assigned, I would leave `major` out, because this tool should only help create minor and patch versions for now, right?

In case there is not one of these labels applied, we should point this out, that the user of this command can fix this PR

OskarStark commented:
And currently I cannot execute this command for `ecommerce`

OskarStark commented:
as we were generating this changelog from now on, should we add the authors like symfony does?

ping @sonata-project/contributors

greg0ire commented:
The authors would be a really really nice touch!

jordisala1991 commented:
Give me the format of authors, and I will add it.

jordisala1991 commented:
I already added The pull request associated with every line of the changelog. But agree that having the author is nice too

jordisala1991 commented:
> This is not IRRELEVANT, there is only the require review label. We should verify, that at least one of docs, pedantic, patch, minor is assigned, I would leave major out, because this tool should only help create minor and patch versions for now, right?

Those red labels are not suposed to be there, but I agree that is will be nice to have an alert about not having any of those labels.

greg0ire commented:
`- my change (@author)` would be best IMO (so, using the github handle, `@` included, not sure why sf does not include it)

jordisala1991 commented:
https://github.com/symfony/symfony/blob/master/CHANGELOG-3.0.md

Not sure if github handles `@` , because it does not handle `#`

greg0ire commented:
Hum right… that's weird…

OskarStark commented:
What about separating the stuff like this

## `CHANGELOG.md`
```
## [3.1.3](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/compare/3.1.2...3.1.3) - 2017-03-22
### Fixed
- [#656] Fixed foo (OskarStark)
- [#655] Fixed bar (greg0ire)
- [#654] Fixed baz (jordisala1991)
```

## For the release tab:
Meaning this one: https://github.com/sonata-project/SonataAdminBundle/releases

This should respect the user handle
```
## [3.1.3](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/compare/3.1.2...3.1.3) - 2017-03-22
### Fixed
- [#656](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/656) Fixed foo (OskarStark)
- [#655](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/655) Fixed bar (greg0ire)
- [#654](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/654) Fixed baz (jordisala1991)
```

Imo it is more convenient to put the ticket number in front of the text, but it really doesn't matter

jordisala1991 commented:
Implemented new changelog format.

Aaaand implemented a message when a PR does not contain any stability labels

jordisala1991 commented:
![captura de pantalla 2017-03-22 a las 20 39 16](https://cloud.githubusercontent.com/assets/1137485/24217307/b1afd6f4-0f3f-11e7-8bec-0f4b086049e8.png)

jordisala1991 commented:
> And currently I cannot execute this command for ecommerce

@OskarStark it does not work with ecommerce because ecommerce has only 3 releases, and none of them is marked as latest release, so github method: `getLatestRelease` returns an exception. IMO this should be fixed on that repo, if 2.0.0 is the latest release, just mark it.

I don't know if only @Soullivaneuh can do that, or we can do it also

jordisala1991 commented:
I think we can do it also, shall we do it? @greg0ire

OskarStark commented:
Yes, please do it @jordisala1991 

btw. good catch 😄

jordisala1991 commented:
IMO we should catch maybe if the lastest tag is not the latest release.

https://github.com/sonata-project/dev-kit/pull/204/files#diff-55aa24e78df8010aea154e9506cab16dR361

If this is correct I will replace irrelevant by pedantic. ok?

jordisala1991 commented:
And I have to restore project argument but in an optional way, because maybe we want to run this command each week and post results on slack programatically

OskarStark commented:
> If this is correct I will replace irrelevant by pedantic. ok?

ok

> And I have to restore project argument but in an optional way

great idea

And maybe we should add another command, which checks all projects and show how many commits we are on top of the latest release, I think this would be more useful information in slack

jordisala1991 commented:
Indeed, then maybe dint add project argument back?

OskarStark commented:
Yes, why not!

jordisala1991 commented:
YAGNI reason :trollface:

core23 commented:
Maybe we could add an auto-tweet to this command to?

jordisala1991 commented:
> Maybe we could add an auto-tweet to this command to?

That could be nice but for the second iteration of this command. Let's keep this PR for just the information about the release: Pull requests, changelog, release number, build status...

jordisala1991 commented:
I changed the way we retrieve pull request, not it is more efficient, and only retrieve those pull requests already merged and > currentRelease date.

DashboardBundle fails because for some reason github returns a different kind of object when finding release >.<

jordisala1991 commented:
Fixed, it was because there is only one branch on DashboardBundle, I changed a bit the code to check if there is only one branch.

jordisala1991 commented:
> For future problems we should catch the exception and point out, that no release is marked as latest release an where the maintainer can do this, WDYT?

Implemented. Maybe we should remove it once all project have an stable release.

> IMO we should catch maybe if the lastest tag is not the latest release.

Super hard to implement, or atleast in a good way. Let's ignore this for now... it should not happen

jordisala1991 commented:
@OskarStark found a bug related to parsing `changelog`. That's why I improved that part.

Now the regex will be able to find the changelog even if it is not using ### for titles, or if there are whitespaces between ``` and `markdown` or after it.

The only requirements are, `## Changelog` and a markdown section with an ending (```).

jordisala1991 commented:
Not sure if i have to change something, or it is working ATM and waiting for @Soullivaneuh review

OskarStark commented:
@Soullivaneuh can you please merge this?

OskarStark commented:
IMO this PR was a big amount of work and @jordisala1991 did it great!

This PR was opened 3 Months ago and still no merge.
@Soullivaneuh what is still todo?
@jordisala1991 are you willing to write tests or could we do this in a separate step.

It would be great to get a tool which provides as much as it can help to create a new release...

jordisala1991 commented:
I can write some tests, but I will also like to know the current state of this PR first. (And IMO it can be done in another step, because this will delay a lot this PR, it won't be easy to test everything)

OskarStark commented:
We should merge this

jordisala1991 commented:
Soon this will be a year old, 🎉

greg0ire commented:
That's cute! https://github.com/tpope/vim-pathogen/pull/142

kunicmarko20 commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,318 @@
+    );
+
+    /**
+     * {@inheritdoc}
```
Nice comment :trollface:

jordisala1991 replied:
Don't worry, I will die before this gets merged :trollface:

kunicmarko20 commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,318 @@
+        $branch = count($branches) > 1 ? next($branches) : current($branches);
+
+        try {
+            $package = $this->packagistClient->get(static::PACKAGIST_GROUP.'/'.$project);
```
why `static::` and not self?

core23 commented:
Can we finally merge this, if this is still an open topic...

greg0ire commented:
> Note: No request change but I will handle the rest manually later.

Does somebody know what this means?

jordisala1991 commented:
No idea, does the PR is still useful?

greg0ire commented:
I will try it for next release and see how it compares to the bash script I wrote, maybe we will end up with a mix of both :P

greg0ire commented on src/Console/Command/ReleaseCommand.php:
```diff
@@ -0,0 +1,318 @@
+                $repositoryName
+            );
+        } catch (ExceptionInterface $e) {
+            throw new RuntimeException('Current release not found');
```
```suggestion
            throw new RuntimeException('Current release not found', 0, $e);
```

jordisala1991 replied:
I removed this exception to help debuggin problems, just like the other one

jordisala1991 commented:
I rebased the PR, squashed all commits, and fixed some of the reviews you have made here.
I can't merge your PR @greg0ire it will need a rebase too. A lot of things changed here.

For the command... IMO we should go with a similar strategy than Symfony uses. They have several commands, but the most important ones are (seen in a Symfony meetup): 

1. Prepare the release: it compiles all the changelog and automatically creates a PR with it.
2. Release: once that PR is merged they have another command that automatically creates the tag, creates the release and attach the changelog.

This PR is going for the first step, we still missing the automatic PR creation but it can be done later too. Once we are good for the first step, we can go for the second to automate the process of creating the tag and the release

greg0ire commented:
I agree, let's not try to do everything at once.

greg0ire commented:
Thanks @jordisala1991 !

jordisala1991 commented:
Thank you :) I will start thinking of next steps
